js/src/common/utils/ItemList.ts

```diff
@@ -81,51 +172,129 @@ export default class ItemList {
-        this.items[i] = items.items[i];
+  merge<K>(otherList: ItemList<K>): ItemList<T | K> {
+    Object.keys(otherList._items).forEach((key) => {
+      const val = otherList._items[key];
+
+      if (val instanceof Item) {
+        (this as ItemList<T | K>)._items[key] = val;
       }
-    }
+    });
 
     return this;
   }
 
   /**
-   * Convert the list into an array of item content arranged by priority. Each
-   * item's content will be assigned an `itemName` property equal to the item's
-   * unique key.
+   * Convert the list into an array of item content arranged by priority.
+   *
+   * This **does not** preserve the original types of primitives and proxies
+   * all content values to make `itemName` accessible on them.
+   *
+   * **NOTE:** If your ItemList holds primitive types (such as numbers, booleans
+   * or strings), these will be converted to their object counterparts if you do
+   * not provide `true` to this function.
+   *
+   * **NOTE:** Modifying any objects in the final array may also update the
+   * content of the original ItemList.
+   *
+   * @param keepPrimitives Converts item content to objects and sets the
+   * `itemName` property on them.
+   *
+   * @see https://github.com/flarum/core/issues/3030
+   */
+  toArray(keepPrimitives?: false): (T & { itemName: string })[];
+  /**
+   * Convert the list into an array of item content arranged by priority.
+   *
+   * Content values that are already objects will be proxied and have
+   * `itemName` accessible on them. Primitive values will not have the
+   * `itemName` property accessible.
+   *
+   * **NOTE:** Modifying any objects in the final array may also update the
+   * content of the original ItemList.
+   *
+   * @param keepPrimitives Converts item content to objects and sets the
+   * `itemName` property on them.
    */
-  toArray(): any[] {
-    const items: Item[] = [];
+  toArray(keepPrimitives: true): (T extends object ? T & Readonly<{ itemName: string }> : T)[];
 
-    for (const i in this.items) {
-      if (this.items.hasOwnProperty(i) && this.items[i] instanceof Item) {
-        this.items[i].content = Object(this.items[i].content);
+  toArray(keepPrimitives: boolean = false): T[] | (T & Readonly<{ itemName: string }>)[] {
+    const items: Item<T>[] = Object.keys(this._items).map((key, i) => {
+      const item = this._items[key];
 
-        this.items[i].content.itemName = i;
-        items.push(this.items[i]);
-        this.items[i].key = items.length;
+      if (!keepPrimitives || isObject(item.content)) {
+        // Convert content to object, then proxy it
+        return {
+          ...item,
+          content: this.createItemContentProxy(isObject(item.content) ? item.content : Object(item.content), key),
+        };
+      } else {
+        // ...otherwise just return a clone of the item.
+        return { ...item };
       }
-    }
+    });
+
+    return items.sort((a, b) => b.priority - a.priority).map((item) => item.content);
```

Comment:
Why remove the check that returns items in order of being added? Or is that already given?

Reply:
We wanted to remove `key` from the Item class, which was used internally for this purpose.

`sort` should default to the order already present if `0` is returned due to the nature of its sort algorithm.

In the below test, 'b' and 'd' both have a priority (`p`) of 3. 'b' is present before 'd', in the initial declaration, and still is after the sort is complete.

![image](https://user-images.githubusercontent.com/7406822/132026785-f4fcb413-7683-410e-9073-0f3f5466140c.png)